### PR TITLE
Move and re-use persist_extra_files

### PR DESCRIPTION
**Pull request by mvdbeek**
This is in preparation of improving the directory datatype.

#BIOHACKEU23

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/job_execution/output_collect.py [hunk 1]
```diff
@@ -56,6 +60,9 @@
     unicodify,
 )
 
+if TYPE_CHECKING:
+    from galaxy.model import DatasetInstance
```
Comment (nsoranzo) [inline lib/galaxy/job_execution/output_collect.py]: We import a bunch of other classes from `galaxy.model` above, so probably we can import `DatasetInstance` there as well, if you like.

File: lib/galaxy/objectstore/__init__.py [hunk 2]
```diff
@@ -1577,3 +1585,27 @@ def set_dataset_object_store_id(self, dataset, require_shareable=True):
         except ObjectInvalid:
             raise Exception("Unable to create output dataset: object store is full")
         self.object_store_id = dataset.object_store_id  # these will be the same thing after the first output
+
+
+def persist_extra_files(object_store: ObjectStore, src_extra_files_path: str, primary_data: "DatasetInstance"):
+    if src_extra_files_path and os.path.exists(src_extra_files_path):
```
Comment (nsoranzo) [inline lib/galaxy/objectstore/__init__.py]: I don't think `src_extra_files_path` can be an empty string.

```suggestion
    if os.path.exists(src_extra_files_path):
```

**Comment by jmchilton** [comment]
What an obviously solid and useful fix. I really love this.

**Review by jmchilton — APPROVED** [review]

**Comment by mvdbeek** [comment]
just a sec though, the integration test failure is real.

File: lib/galaxy/objectstore/__init__.py (hunk 2 above)
Comment (mvdbeek) [reply lib/galaxy/objectstore/__init__.py]: It certainly could be in the past if the extra files dir hadn't been created yet. Not sure I'm ready remove the safe-guard ?

File: lib/galaxy/job_execution/output_collect.py (hunk 1 above)
Comment (mvdbeek) [reply lib/galaxy/job_execution/output_collect.py]: Moved it back to the top

File: lib/galaxy/objectstore/__init__.py [hunk 3]
```diff
@@ -1577,3 +1585,25 @@ def set_dataset_object_store_id(self, dataset, require_shareable=True):
         except ObjectInvalid:
             raise Exception("Unable to create output dataset: object store is full")
         self.object_store_id = dataset.object_store_id  # these will be the same thing after the first output
+
+
+def persist_extra_files(object_store: ObjectStore, src_extra_files_path: str, primary_data: "DatasetInstance"):
```
Comment (nsoranzo) [inline lib/galaxy/objectstore/__init__.py]:
```suggestion
def persist_extra_files(object_store: ObjectStore, src_extra_files_path: str, primary_data: "DatasetInstance") -> None:
```

File: lib/galaxy/job_execution/output_collect.py [hunk 4]
```diff
@@ -723,9 +727,11 @@ def default_exit_code_file(files_dir, id_tag):
     return os.path.join(files_dir, f"galaxy_{id_tag}.ec")
 
 
-def collect_extra_files(object_store, dataset, job_working_directory):
+def collect_extra_files(object_store: ObjectStore, dataset: "DatasetInstance", job_working_directory: str):
```
Comment (nsoranzo) [inline lib/galaxy/job_execution/output_collect.py]:
```suggestion
def collect_extra_files(object_store: ObjectStore, dataset: "DatasetInstance", job_working_directory: str) -> None:
```

**Review by nsoranzo — APPROVED** [review]

**Pull request merged**